Feat: Update Companies + Job Listing

Okumans opened the pull request:
This pull request introduces significant changes to the codebase, primarily focusing on the addition of job listing functionality, updates to company-related operations, and improvements to the testing setup. The most important changes include adding routes and controllers for job listings, modifying company-related controllers to integrate job listings, and setting up an in-memory MongoDB for testing.

### Job Listings Functionality:

* Added routes and controllers for job listings, including `getJobListing`, `getJobListings`, `getJobListingByCompany`, `createJobListing`, `updateJobListing`, and `deleteJobListing` in `src/controllers/jobListing.ts` and `src/routes/jobListing.ts`. [[1]](diffhunk://#diff-689b852ef9cad322472a355150a5c5e32e7e0c3acb0e7755fa7c450af4a47cf9R1-R165) [[2]](diffhunk://#diff-90287b7ab83b57e68118988c041ac4f392beecde03d7019286a51325840615dbR1-R29)
* Updated `CompanyModel` to include a virtual field for job listings and added references to job listings in the `Company` schema in `src/models/Company.ts`. [[1]](diffhunk://#diff-0fbc384876c39162815c982803faabfee61d9df29d4f4483fe458d8959e5fcfeL3-R3) [[2]](diffhunk://#diff-0fbc384876c39162815c982803faabfee61d9df29d4f4483fe458d8959e5fcfeL33-R52)
* Modified `InterviewSessionModel` to reference job listings instead of companies in `src/models/InterviewSession.ts`.
* Added a new `JobListingModel` schema in `src/models/JobListing.ts`.

### Company-related Operations:

* Updated company controllers to integrate job listings, including changes to `getCompanies`, `getCompany`, `createCompany`, `updateCompany`, and `deleteCompany` in `src/controllers/companies.ts`. [[1]](diffhunk://#diff-4a7538d5f1b43a82e5b029b2f48df2636a4b8f32a639f66680376a4b28e7be88L22-R24) [[2]](diffhunk://#diff-4a7538d5f1b43a82e5b029b2f48df2636a4b8f32a639f66680376a4b28e7be88L57-R63) [[3]](diffhunk://#diff-4a7538d5f1b43a82e5b029b2f48df2636a4b8f32a639f66680376a4b28e7be88R87-R90) [[4]](diffhunk://#diff-4a7538d5f1b43a82e5b029b2f48df2636a4b8f32a639f66680376a4b28e7be88L110-R116) [[5]](diffhunk://#diff-4a7538d5f1b43a82e5b029b2f48df2636a4b8f32a639f66680376a4b28e7be88L136-R171)
* Modified routes to include job listing endpoints and updated authorization rules in `src/routes/companies.ts`. [[1]](diffhunk://#diff-f6522086f03fad7f13cee263da60f0f690b308ed6f4fbe1e8c1f2d186696d822R9) [[2]](diffhunk://#diff-f6522086f03fad7f13cee263da60f0f690b308ed6f4fbe1e8c1f2d186696d822L17-R39)

### Testing Improvements:

* Set up an in-memory MongoDB for testing purposes in `tests/setup.ts`. This includes starting and stopping the MongoDB memory server and connecting to it before tests run.

### Other Changes:

* Added `jest` as a dependency in `package.json` for testing.
* Minor updates to error handling in `src/middleware/errorHandler.ts`.
* Added a new route for job listings in `src/app.ts`. [[1]](diffhunk://#diff-841254fe75488c1bd4cd7f68f00b4be0e48dcfbc4a16b45847b68295e0e3b27bR14) [[2]](diffhunk://#diff-841254fe75488c1bd4cd7f68f00b4be0e48dcfbc4a16b45847b68295e0e3b27bR37)

6733275621 committed:
feat: add company to User schema

6733275621 committed:
feat: add JobListing model

6733275621 committed:
fix: add require in JobListing schema

6733275621 committed:
feat: change company to jobListing in Session schema

6733275621 committed:
feat: add jest

Kentakoong committed:
chore: remove package lock

6733275621 committed:
feat: add jobListing and owner to Company schema

6733275621 committed:
feat: add jobListing controller

6733275621 committed:
fix: change company populate from session to jobListing

6733275621 committed:
feat: add jobListing route

6733275621 committed:
feat: add jobListingRouter to server

6733275621 committed:
fix: api joblisting -> job-listing

6733275621 committed:
fix: jobListingSchema add image, description String -> Mixed, remove some field

6733275621 committed:
feat: add getJobListingByCompany, remove getJobListings

6733275621 committed:
fix: remove jobListing array from Company model

6733275621 committed:
fix: remove company from session model

6733275621 committed:
feat: add getInterviewSessionByJobListing

6733275621 committed:
fix: change jobListing model description to string

6733275621 committed:
fix: config.env to root

6733275621 committed:
fix: remove pnpm dependency

6733275621 committed:
fix: models interface objectId to mongoose.Types.ObjectId

Okumans committed:
feat: add Nullable and Combine for Global type definitions

Okumans committed:
feat: add logo field to Company model and type definition

- Added a new 'logo' field to the Company schema in Company.ts to store the company's logo URL.
- Updated the Company interface in Company.d.ts to include the 'logo' property as a nullable string.

Okumans committed:
fix: update phone number validation regex and message in Company model

6733275621 committed:
fix: add back missing functions

6733275621 committed:
feat: add company id to user when creating company

Okumans committed:
refactor: use mongoose.Types.ObjectId for Company, Session, Listing, and User in Document

Okumans committed:
fix: change 400 to 404 for companies and joblistings not found error responses

Okumans committed:
feat: cascade delete joblistings and interviewSessions in deleteCompany

Okumans committed:
feat: enhance test setup with MongoDB Memory Server integration and improved error handling

Okumans committed:
fix: update error response structure to use 'message' instead of 'error'

Kentakoong commented on src/app.ts:
```diff
@@ -11,6 +11,7 @@ import { authRouter } from "@/routes/auth";
 import { companiesRouter } from "@/routes/companies";
 import { interviewSessionsRouter } from "@/routes/interviewSessions";
 import { usersRouter } from "@/routes/users";
+import { jobListingsRouter } from "./routes/jobListing";
```
use absolute paths

Kentakoong commented on src/controllers/interviewSessions.ts:
```diff
@@ -222,6 +222,8 @@ export const deleteInterviewSession = async (
 };
 
+
+// Unuseable now. Remove?
```
yeah just remove it if you don't think we use it

Kentakoong commented on src/middleware/errorHandler.ts:
careful about this, if this is changed we need to change params in frontend, really not recommended to do this

Okumans replied:
all `message` instance related with error is now renamed to `error`

Okumans committed:
refactor: refactor error responses and improve test setup

Updated all controller error responses to use the "error" field instead
of "message"

Modified Jest configuration to include setup files and adjust
moduleNameMapper

Improve setup of in-memory MongoDB

Change the test file extension from `.ts` to `.test.ts`

Improved error handling in middleware for consistency and clarity

Okumans committed:
fix: use absolute paths instead of relative for jobListingsRouter in app.ts

Okumans committed:
feat: renamed `sessions` to `interviewSessions` in jobListing.ts for more clarity

Okumans committed:
refactor: remove unused `getInterviewSessionsByCompany` and related routes from companies

Okumans committed:
refactor: renamed jobListing routes and controllers to jobListings for consistency

Okumans committed:
refactor: rename function `getJobListingByCompany` to getJobListingsByCompany for consistency

Okumans committed:
refactor: changed virtual field name from 'jobListing' to 'jobListings' in Company model

Okumans committed:
fix: update deleteCompany response to return empty data object

Kentakoong commented on dist/tsconfig.tsbuildinfo:
ignore this file, use git rm to remove it from git

Kentakoong commented on src/middleware/errorHandler.ts:
```diff
@@ -11,26 +11,26 @@ export const errorHandler = (
     res.status(statusCode).json({
         success: false,
-        error: message,
+        message: error,
```
why is it still returning as message: but not error:

I want to recieve the error message as error

Okumans replied:
my bad, forgot 😅

Okumans committed:
fix: add tsconfig.tsbuildinfo to .gitignore

Okumans committed:
fix: change error response key from 'message' to 'error' in errorHandler
